app/src/test/kotlin/compiler/intermediate_form/ControlFlowGraphBuilderTest.kt

```diff
@@ -0,0 +1,144 @@
+        entryTreeRoot = entryNode,
+        unconditionalLinks = referenceHashMapOf(entryNode to secondNode),
+        conditionalTrueLinks = referenceHashMapOf(secondNode to conditionalTrueNode),
+        conditionalFalseLinks = referenceHashMapOf(secondNode to conditionalFalseNode)
+    )
+
+    @Test
+    fun `test pass entryTreeRoot in constructor`() {
+        val cfg = ControlFlowGraphBuilder(entryNode).build()
+
+        assertEquals(cfg.entryTreeRoot, entryNode)
+        assertEquals(cfg.treeRoots, listOf(entryNode))
+        assertEquals(cfg.unconditionalLinks, referenceHashMapOf())
+        assertEquals(cfg.conditionalFalseLinks, referenceHashMapOf())
+        assertEquals(cfg.conditionalTrueLinks, referenceHashMapOf())
+    }
+
+    @Test
+    fun `test makeRoot`() {
+        val cfgBuilder = ControlFlowGraphBuilder()
+        cfgBuilder.makeRoot(entryNode)
+
+        val cfg = cfgBuilder.build()
+        assertEquals(ControlFlowGraphBuilder(entryNode).build(), cfg)
+    }
+
+    @Test
+    fun `test makeRoot when already specified`() {
+        val cfgBuilder = ControlFlowGraphBuilder(entryNode)
+        assertFailsWith<IncorrectControlFlowGraphError> { cfgBuilder.makeRoot(entryNode) }
+    }
+
+    @Test
+    fun `test addLink from null sets entryTreeRoot`() {
+        val cfgBuilder = ControlFlowGraphBuilder()
+        cfgBuilder.addLink(null, entryNode)
+        cfgBuilder.addLink(Pair(entryNode, CFGLinkType.UNCONDITIONAL), secondNode)
+        cfgBuilder.addLink(Pair(secondNode, CFGLinkType.CONDITIONAL_TRUE), conditionalTrueNode)
+        cfgBuilder.addLink(Pair(secondNode, CFGLinkType.CONDITIONAL_FALSE), conditionalFalseNode)
+
+        val cfg = cfgBuilder.build()
+        assertEquals(expectedCFG, cfg)
+    }
+
+    @Test
+    fun `test addLinkFromAllFinalRoots`() {
+        val cfgBuilder = ControlFlowGraphBuilder(entryNode)
+        cfgBuilder.addLinkFromAllFinalRoots(CFGLinkType.UNCONDITIONAL, secondNode)
+        cfgBuilder.addLinkFromAllFinalRoots(CFGLinkType.CONDITIONAL_TRUE, conditionalTrueNode)
+        cfgBuilder.addLink(Pair(secondNode, CFGLinkType.CONDITIONAL_FALSE), conditionalFalseNode)
+
+        val cfg = cfgBuilder.build()
+        assertEquals(expectedCFG, cfg)
+    }
+
+    @Test
+    fun `test addLinkFromAllFinalRoots when no entryTreeRoot`() {
+        val cfgBuilder = ControlFlowGraphBuilder()
+        cfgBuilder.addLinkFromAllFinalRoots(CFGLinkType.UNCONDITIONAL, entryNode)
+        assertEquals(ControlFlowGraphBuilder(entryNode).build(), cfgBuilder.build())
+    }
+
+    @Test
+    fun `test addAllFrom`() {
+        val cfgBuilder = ControlFlowGraphBuilder()
+        cfgBuilder.addAllFrom(expectedCFG)
+
+        val cfg = cfgBuilder.build()
+        assertEquals(expectedCFG, cfg)
+    }
+
+    @Test
+    fun `test addAllFrom without modifying entryTreeRoot`() {
+        val cfgBuilder = ControlFlowGraphBuilder(entryNode)
+
+        val remainingTwoNodes = ControlFlowGraphBuilder(secondNode)
```

Comment:
Actually, this CFG consists of **three** nodes.

Reply:
https://www.youtube.com/watch?v=iXpxYn2uEEg&ab_channel=EvilBrotherThaddeus